presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -2559,6 +2603,27 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
         return toCompletableFuture(stagingFileCommitter.commitFiles(session, handle.getSchemaName(), handle.getTableName(), getPartitionUpdates(fragments)));
     }
 
+    private HivePartitioningHandle createHivePartitioningHandle(
+            int bucketCount,
+            HiveBucketHandle hiveBucketHandle,
+            Optional<HiveBucketProperty> bucketProperty,
```

Comment:
I initially use non-optional, but there was failed test due to empty `HiveBucketProperty`

Reply:
In what case the `HiveBucketProperty` is missing?

Reply:
Here are some examples: https://api.travis-ci.com/v3/job/347805400/log.txt,   `HiveBucketProperty` is optional inside `Storage` class

Reply:
Okay, now I understand. The `bucketProperty` will be missing only if the table is "virtually bucketed". Let's have a special case for that to make it explicit, otherwise it is very confusing. 

Honestly i don't really like this method. It is very confusing. It is very hard to understand in what cases  `bucketProperty` is present, why `bucketCount`, `hiveBucketHandle` and `maxCompatibleBucketCount` may have different bucket count, etc. I would recommend to inline this method. It will result in a couple of lines of duplicated code, but it should make things much clearer.

Reply:
The constructor is inline now.